x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts

```diff
@@ -0,0 +1,330 @@
+import { getRuleData } from './get_rule_executor_data';
+
+type LifecycleAlertService<
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> = (alert: {
+  id: string;
+  fields: Record<string, unknown>;
+}) => AlertInstance<InstanceState, InstanceContext, ActionGroupIds>;
+
+export interface LifecycleAlertServices<
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> {
+  alertWithLifecycle: LifecycleAlertService<InstanceState, InstanceContext, ActionGroupIds>;
+}
+
+export type LifecycleRuleExecutor<
+  Params extends AlertTypeParams = never,
+  State extends AlertTypeState = never,
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> = (
+  options: AlertExecutorOptionsWithExtraServices<
+    Params,
+    State,
+    InstanceState,
+    InstanceContext,
+    ActionGroupIds,
+    LifecycleAlertServices<InstanceState, InstanceContext, ActionGroupIds>
+  >
+) => Promise<State | void>;
+
+const trackedAlertStateRt = rt.type({
+  alertId: rt.string,
+  alertUuid: rt.string,
+  started: rt.string,
+});
+
+export type TrackedLifecycleAlertState = rt.TypeOf<typeof trackedAlertStateRt>;
+
+const alertTypeStateRt = <State extends AlertTypeState>() =>
+  rt.record(rt.string, rt.unknown) as rt.Type<State, State, unknown>;
+
+const wrappedStateRt = <State extends AlertTypeState>() =>
+  rt.type({
+    wrapped: alertTypeStateRt<State>(),
+    trackedAlerts: rt.record(rt.string, trackedAlertStateRt),
+  });
+
+/**
+ * This is redefined instead of derived from above `wrappedStateRt` because
+ * there's no easy way to instantiate generic values such as the runtime type
+ * factory function.
+ */
+export type WrappedLifecycleRuleState<State extends AlertTypeState> = AlertTypeState & {
+  wrapped: State | void;
+  trackedAlerts: Record<string, TrackedLifecycleAlertState>;
+};
+
+export const createLifecycleExecutor = (logger: Logger, ruleDataClient: RuleDataClient) => <
+  Params extends AlertTypeParams = never,
+  State extends AlertTypeState = never,
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+>(
+  wrappedExecutor: LifecycleRuleExecutor<
+    Params,
+    State,
+    InstanceState,
+    InstanceContext,
+    ActionGroupIds
+  >
+) => async (
+  options: AlertExecutorOptions<
+    Params,
+    WrappedLifecycleRuleState<State>,
+    InstanceState,
+    InstanceContext,
+    ActionGroupIds
+  >
+): Promise<WrappedLifecycleRuleState<State>> => {
+  const {
+    rule,
+    services: { alertInstanceFactory },
+    state: previousState,
+  } = options;
+
+  const ruleExecutorData = getRuleData(options);
+
+  const state = getOrElse(
```

Comment:
@weltenwort I see you refactored this part. My understanding is that this refactoring was done to accommodate generic values.  I don't fully understand the `isLeft` part in the initial code, so can you verify that the refactored code has the same result as before? 

```
const decodedState = wrappedStateRt.decode(previousState);

      const state = isLeft(decodedState)
        ? {
            wrapped: previousState,
            trackedAlerts: {},
          }
        : decodedState.right;
```

Reply:
`isLeft` will return `true` if the value could not be decoded ("left"). In that case, we wrap the state returned by the wrapped executor. If it can be successfully decoded ("right"), we know that it is a state managed by the wrapper, so we return it as-is.

Reply:
@dgieselaar Thanks for clarification!